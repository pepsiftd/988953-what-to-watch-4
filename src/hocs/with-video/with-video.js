import React, {PureComponent, createRef} from 'react';
import PropTypes from 'prop-types';

const withVideo = (Component) => {
  class WithVideo extends PureComponent {
    constructor(props) {
      super(props);
      console.log(props);
      this._videoSettings = props.videoSettings;

      this._videoRef = createRef();

      this.state = {
        isMute: this._videoSettings.isMute,
        isActive: this._videoSettings.isActive,
      };
    }

    componentDidMount() {
      const {src, poster, width, height} = this._videoSettings;
      const {isMute} = this.state;
      const video = this._videoRef.current;

      video.src = src;
      video.poster = poster;
      video.muted = isMute;
      video.width = width;
      video.height = height;
    }

    componentWillUnmount() {
      const video = this._videoRef.current;

      video.src = ``;
      video.poster = ``;
      video.onplay = null;
      video.onpause = null;
      video.muted = null;
      video.width = null;
      video.height = null;
    }

    componentDidUpdate() {
      const video = this._videoRef.current;

      if (this._videoSettings.isActive) {
        video.play();
      } else {
        video.load();
      }
    }

    render() {

      return (
        <Component
          {...this.props}
        >
          <video
            ref={this._videoRef}
          />
        </Component>
      );
    }
  }

  WithVideo.propTypes = {
    videoSettings: PropTypes.shape({
      src: PropTypes.string.isRequired,
      poster: PropTypes.string.isRequired,
      isActive: PropTypes.bool.isRequired,
      isMute: PropTypes.bool.isRequired,
      width: PropTypes.string.isRequired,
      height: PropTypes.string.isRequired,
    })
  };

  return WithVideo;
};

export {withVideo};
