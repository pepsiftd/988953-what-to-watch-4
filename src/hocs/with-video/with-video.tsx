import React, {createRef} from 'react';
import {Subtract} from 'utility-types';

interface InjectingProps {
  children: React.ReactNode;
}

const withVideo: (Component: React.ComponentClass | React.FunctionComponent) => React.ComponentClass = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithVideo extends React.Component<T> {
    private videoRef: React.RefObject<HTMLVideoElement>;
    private videoSettings: {
      src: string;
      poster: string;
      width: number;
      height: number;
      isMute: boolean;
    };

    constructor(props) {
      super(props);
      this.videoSettings = props.videoSettings;

      this.videoRef = createRef();
    }

    shouldComponentUpdate(nextProps) {
      return this.props.isActive !== nextProps.isActive;
    }

    componentDidMount() {
      const {src, poster, width, height, isMute} = this.videoSettings;
      const video = this.videoRef.current;

      video.src = src;
      video.poster = poster;
      video.muted = isMute;
      video.width = width;
      video.height = height;
    }

    componentWillUnmount() {
      const video = this.videoRef.current;

      video.src = ``;
      video.poster = ``;
      video.onplay = null;
      video.onpause = null;
      video.muted = null;
      video.width = null;
      video.height = null;
    }

    componentDidUpdate() {
      const video = this.videoRef.current;

      if (this.props.isActive) {
        video.play();
      } else {
        video.load();
      }
    }

    render() {

      return (
        <Component
          {...this.props}
        >
          <video
            ref={this.videoRef}
          />
        </Component>
      );
    }
  }

  return WithVideo;
};

export {withVideo};
