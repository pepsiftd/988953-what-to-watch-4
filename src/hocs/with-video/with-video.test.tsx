import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';

import {withVideo} from './with-video';

const MockComponent = ({children}) => {
  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired
};

const MockComponentWithVideo = withVideo(MockComponent);

it(`withVideo should render correctly`, () => {
  const tree = renderer.create(
      <MockComponentWithVideo
        videoSettings={{
          src: `source.mp4`,
          poster: `image.jpg`,
          isMute: true,
          width: `280`,
          height: `175`,
        }}
        isActive={true}
      />, {
        createNodeMock() {
          return {};
        }
      }
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
