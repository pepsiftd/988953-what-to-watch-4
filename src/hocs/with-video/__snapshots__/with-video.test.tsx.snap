// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withVideo should render correctly 1`] = `
<div>
  <video />
</div>
`;
